File: go/hw_test.go
// Copyright 2016, yubo.
// All rights reserved.

package hw

import (
	"testing"
)

func TestHw(t *testing.T) {
	out := Hw()
	if out != "hello,world" {
	}
	if got, want := Hw(), `hello,world`; got != want {
		t.Errorf("got %q want %q", got, want)
	}
}

func TestHwI(t *testing.T) {
	tests := []struct {
		x    interface{}
		want string
	}{
		{bool(true), `hello,world bool true`},
		{float32(1.01), `hello,world float32 1.01`},
		{int(42), `hello,world int 42`},
	}
	for _, e := range tests {
		if got, want := HwI(e.x), e.want; got != want {
			t.Errorf("got %q want %q", got, want)
		}
	}

}

File: c/hw_test.c
// Copyright 2016, yubo.
// All rights reserved.

#include "hw.h"

#define BSIZE 64

char buff[BSIZE+1];

static int testHw(void){
	const char *want = "hello,world";

	hw(buff, BSIZE);
	if (strncmp(want, buff, BSIZE)){
		printf("got \"%s\" want \"%s\"\n", buff, want);
		return -1;
	}
	return 0;
}

static int testHwi(void){
	size_t i;

	struct {
		int n;
		char want[BSIZE+1];
	} tests[] = {
		{5, "hello,world small"},
		{50, "hello,world big"},
	};

	for(i = 0; i < sizeof(tests)/sizeof(tests[0]); i++){
		hwi(tests[i].n, buff, BSIZE);
		if (strncmp(tests[i].want, buff, BSIZE)){
			printf("got \"%s\" want \"%s\"\n", buff, tests[i].want);
			return -1;
		}
	}
	return 0;
}

int main(){

	if(testHw())
		return -1;

	if(testHwi())
		return -1;

	printf("PASS\n");
	return 0;
}
